### PR TITLE
[ISSUE #549] Unify request handler code style of tcp and http protocol

### LINKED ISSUE #549: [Enhancement]Unify request handler code style of tcp and http protocol
### Is your enhancement request related to a problem? Please describe
Http protocol handle with processors but Tcp protocol handle with tasks, it is needed to unify them into a same code style, it would be better for futher protocol extendtion.

### PR DESCRIPTION
**Pull request by silentchildh**
Fixes #549 

### Motivation

- Unified TCP and HTTP code styles
- Optimize code encapsulation and reduce coupling
- More careful subcontracting and naming enhances readability

#### boot 
Firstly, I divided the server into three layers:
1. The most basic layer is `AbstractRemotingServer` . It leverages Netty for the most basic communication server.
2. Then there are the protocol-layer servers that enable the runtime module to run simply, such as `AbstractTcpServer` and `AbstractHttpServer`.
3. Finally, there are `EventMeshXxxServer` that integrate multiple capabilities (typically a manager that assembles the capabilities to enhance the server) , such as `EventMeshTCPServer` and 'EventMeshHTTPServer`.

In my opinion, the most important part of the runtime module is the "processor" . The main component of the second-tier server is the "processor" . That is, declare the "processor" required for registration in the Layer 2 server `AbstractXXXServer` to build the layer server.

In layer 3 server, I prefer pluggable plug-in design. That is, in the process of extending the "processor" , if there is a function that requires other modules, add the manager of the corresponding function in this layer. The processor then retrieves the corresponding manager through the `EventMeshXxxServer` instance.

For example :
- To better manage event/message production and consumption in the runtime module, `ProducerManager` and `ConsumerManager` are provided.
- `Registry` is provided for cluster deployment of eventmesh servers.
- The rest of the modules, such as 'security' and 'metric' , can be inserted into the corresponding manager in the tier 3 server.

#### core

Here is my understanding of the refactored core package and what it takes to implement a server based on `EventMeshHTTPServer`.
- Producer is used to upload messages from the client to the message storage.
- Consumer is used to download messages from the message storage to the client.
- The retry mechanism is needed to ensure the security and reliability of messages, whether they are uploaded or downloaded.
- The server-side operation is based on the processor implementation.

### Modifications

boot:
- So I implemented TCP's second-tier server, `AbstractTcpServer` . I've used all of Netty's required handlers as `AbstracTcpServer` internal classes to better manage this tier of servers. 
- And registered "processor" in the third-tier server juts like `EventMeshHTTPServer`
- I think the creation and use of threads is actually the basis for the server to run, so I put it on the second-tier server and used an interface `ThreadPoolGroup` to normalize it, the implementation classes of this interface can assemble the required thread pool.

core:
- During the refactoring I found that producer and consumer can actually be abstracted out and used by both TCP and HTTP.
- Also related to HTTP are  `SubcriptionManager` and `HttpClientGroupMapping`, which hold data locally on the `EventMeshHTTPServer` about the relationships between clients, producers, and subscription items, so I pulled it out and put it in the local package.
- The original `ClientSessionGroupMapping` and `ClientGroupWrapper` classes in TCP record client, session, producer, consumer, and subscription project relationships locally. Now that I've decoupled it so that `ClientManager` focuses on the relationship between the client and the session, `SessionManager` uses `ProducerMap` to record the relationship between the producer and the session, and `ConsumerMap` to record the relationship between the consumer and the session, `SubscriptionMap` records the relationship of the subscription item to the session.

### Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
- If a feature is not applicable for documentation, explain why?
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Review by Pil0tXia — COMMENTED** [review]
Idea is good.

image

image

I'd like to know more about your decoupled design for nested consumers and producers.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/consumer/ConsumerManager.java [hunk 1]
```diff
@@ -15,106 +15,172 @@
...
 
-    public void shutdown() {
-        eventMeshHTTPServer.getEventBus().unregister(this);
-        for (ConsumerGroupManager consumerGroupManager : consumerTable.values()) {
-            try {
-                consumerGroupManager.shutdown();
-            } catch (Exception ex) {
-                log.error("shutdown consumerGroupManager[{}] err", consumerGroupManager, ex);
+    @Subscribe
+    public void handleConsumerGroupTopicConfUpdateEvent(ConsumerGroupTopicConfUpdateEvent event) {
```
[the remark this answers is not in the document]
Comment (silentchildh) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/consumer/ConsumerManager.java:118]: The two methods with the `@subscribe` annotation are waiting for the `notifyConsumerManager` method to be called and will be received by the two methods when the `EventBus` `post` method is used.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/consumer/push/HTTPPushRequestHandler.java [hunk 2]
```diff
@@ -42,67 +41,67 @@
...
-        this.pushExecutor = eventMeshConsumer.getEventMeshHTTPServer().getPushMsgExecutor();
-        waitingRequests.put(this.eventMeshConsumer.getConsumerGroupConf().getConsumerGroup(), Sets.newConcurrentHashSet());
-        SCHEDULER.scheduleAtFixedRate(this::checkTimeout, 0, 1000, TimeUnit.MILLISECONDS);
-    }
-
     @Override
-    public boolean handle(final HandleMsgContext handleMsgContext) {
-        if (MapUtils.getObject(waitingRequests, handleMsgContext.getConsumerGroup(), Sets.newConcurrentHashSet()).size()
+    public boolean push(final PushRequestContext pushRequestContext) {
+        if (MapUtils.getObject(waitingRequests, pushRequestContext.getConsumerGroup(), Sets.newConcurrentHashSet()).size()
```
[the remark this answers is not in the document]
Comment (silentchildh) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/consumer/push/HTTPPushRequestHandler.java:59]: Yes, I may not have finished all my work. My idea at the time was to abstract away TCP and HTTP from pushing messages to the client, so I wrote a `Pushable` interface first.

**Comment by silentchildh** [comment]
I don't get too involved with the consumer and producer aspects of the HTTP module, I prefer to abstract them so that TCP can reuse it better. In fact, my main focus is to decouple TCP's original `ClientSessionGroupMapping` and `ClientGroupWrapper`.

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/consumer/push/HTTPPushRequestHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Pil0tXia) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/core/protocol/http/consumer/push/HTTPPushRequestHandler.java:59]: OK. You may finish it later~

**Comment by Pil0tXia** [comment]
>Q: I'd like to know more about your decoupled design for nested consumers and producers.

>A: I don't get too involved with the consumer and producer aspects of the HTTP module, I prefer to abstract them so that TCP can reuse it better.

With your nested design, it would seem that the Consumer/Producer in the `core` folder is shared by all protocols, and that the Consumer/Producer in each protocol folder is an implementation of the parent folder. But in practice this doesn't seem to be the case, the Consumer/Producer in the parent and child folders don't have the similar functionality or meaning, and even in `tcp` folder there is no significant referential relationship. I think you could optimise a bit more in this area to improve reusability or readability. 

Other than that, the modifications are fine with me.

**Comment by silentchildh** [comment]
@Pil0tXia 
For now, the top-level consumer and producer in the core package are the respective managers, and the consumer and producer packages under each protocol are some of the data structures needed to correspond to send and push messages.

For example, `SendMessageContext` is used to encapsulate redeliverable information, and `Sessionsender` and `Sessionpusher` are used to send information to the corresponding client. The Manager of both the Consumer and Producer will actually call them.

- For Producer, you need to receive the active push from the client and then process the message through `SendMessageProcessor` and `Sessionsender`.
- In particular, the Consumer does not use `processor` when consuming information.When the Consumer consumes the information, it actually subscribes to the `EventMeshXxxServer` receiving the end-to-end message when the `EventMeshConusmer` is initialized. This is what the original HTTP server did.

For more detailed information, please take a close look at `org. Apache. Eventmesh. Runtime. Core. Consumer`. The init method in the `Eventmeshconsumer` Class

**Comment by silentchildh** [comment]
Because eventmesh is going to use the connect module instead of the original consumer and producer, I'll turn on another PR to avoid changes to consumer and producer. Instead, just optimize the EventMeshServer.

**Pull request closed without merging**